docs/dev/Docker.md

```diff
@@ -0,0 +1,73 @@
+# How to build, publish and run a Apache Drill Docker image
+
+## Prerequisites
+
+To build an Apache Drill docker image, you need to have the following software installed on your system to successfully complete a build. 
+  * Java 8
+  * Maven 3.3.1 or greater
+  * Docker 18 or greater
+
+## Checkout
+
+`git clone https://github.com/apache/drill.git`
+    
+## Build Drill
+```
+$ cd drill
+$ mvn clean install
+```   
+## Build Docker Image
+```
+$ cd distribution
+$ mvn dockerfile:build -Pdocker
+```
+
+## Push Docker Image
+   
+   By default, the image built above is configured to be pushed to [Drill Docker Hub](https://hub.docker.com/r/drill/apache-drill/) to create official Drill Docker images. 
+```   
+$ cd distribution
+$ mvn dockerfile:push -Pdocker
+```    
+  You can configure the repository in pom.xml to point to any private or public container registry, or specify it in your mvn command:
+```  
+$ cd distribution
+$ mvn dockerfile:push -Pdocker -Pdocker.repository=<my_repo>
+```
+## Run Docker Container
+   
+   Running the Docker container should start Drill in embedded mode and connect to Sqlline
+```    
+$ docker run -i -t drill/apache-drill:1.14.0 /bin/bash
```

Comment:
Please add the following clarifications:
1. Detach: When starting this way (without detach), once you exit the sqline, container will be also killed. To persist the container use `--detach`: `docker run -i --detach -t drill/apache-drill:1.14.0 /bin/bash /bin/bash`.
2. Port forwarding: In order to access Drill WebUI you need to open 8047 port: `docker run -i -p 8047:8047 -t drill/apache-drill:1.14.0 /bin/bash`.
3. Name: Suggest optionally to add name to the container (`--name drill`): `docker run --name drill -i -p 8047:8047 -t drill/apache-drill:1.14.0 /bin/bash`.
4. Docker machine: On Windows and Mac default docker machine might not have enough resources.
On my env I had to create new docker machine. Example:
```
# stop default
docker-machine stop default
# add drill-machine
docker-machine create -d virtualbox --virtualbox-cpu-count "2" --virtualbox-memory "16384" drill-machine
# activate drill-machine
eval $(docker-machine env drill-machine)
# check status
docker-machine active
```
5. Querying custom files: To query custom files (which I believe will be the major use case), user needs to mount a volume (https://docs.docker.com/storage/volumes/) and then update file storage plugin to look at that volume. I did not try that but it's worth mentioning and even adding a simple manual.

Reply:
(1) --detach option is useful. Users will then need to use drill-localhost to connect to Sqlline once they connect into the container. Will update the doc.
(2) I don't think it's necessary. Also, it's preferred to use localhost:8047 to connect to Web UI if brought up on Mac or PC.  
(3) Name will be useful - this way we'll not need to provide the container ID during Docker exec. Will add this and update the doc. 
(4) I tried on Docker for Mac - there were no issues. Looking at the doc (https://docs.docker.com/machine/overview/#what-is-docker-machine), it's for older Mac systems which do not support Docker for Mac (probably similarly for PCs). If someone faces an issue, they'll likely face it for any container they try to bring up - independent of Drill. So this is more generic Docker troubleshooting. I'm thinking, we can just have a link to this instead? 
(5) Volumes will be helpful - for now, i'll share a link to this in the doc. Will try it out and include examples in the doc Bridget will add - as it's a more user facing doc (ex: how to query data on your Mac when using Drill Docker container)

Reply:
(1) - Sure, thanks.
(2) - I had to do port forwarding on my machine (Windows 10), it did not work without it, neither  is with localhost, had to use ip + port.
(3) - Sure, thanks
(4) - had an issue on my env Windows 10, you just can mention it for Windows users. Drill is resource consuming, for example, I was able to bring up mysql with default docker machine.
(5) - Sure, thanks

Reply:
Done, please check. 
Regarding (2), i'll update the doc with port forwarding (specific to Windows perhaps). 
For (4), I've mentioned it in the pre-requisites.